src/c_api.cpp

```diff
@@ -364,9 +373,13 @@ class Booster {
  private:
   const Dataset* train_data_;
   std::unique_ptr<Boosting> boosting_;
-  std::unique_ptr<Predictor> single_row_predictor_;
-  PredictFunction single_row_predict_function_;
-  int64_t single_row_num_pred_in_one_row_;
+  std::unique_ptr<Predictor> single_row_predictor_[PREDICTOR_TYPES];
+  bool single_row_predictor_pred_early_stop[PREDICTOR_TYPES];
+  int single_row_predictor_pred_early_stop_freq[PREDICTOR_TYPES];
+  double single_row_predictor_pred_early_stop_margin[PREDICTOR_TYPES];
+  int single_row_predictor_iter_[PREDICTOR_TYPES];
+  PredictFunction single_row_predict_function_[PREDICTOR_TYPES];
+  int64_t single_row_num_pred_in_one_row_[PREDICTOR_TYPES];
```

Comment:
maybe have a new class for the single row prediction? due to there are much more member variables of it.
It will be easier to maintain.

Reply:
I think in this file is okay.